pyrk/th_system.py

```diff
@@ -104,3 +107,262 @@ def custom(self, t_b, t_env, res):
         num = (t_b-t_env)
         denom = res.to(units.kelvin/units.watt)
         return num/denom
+
+
+class THSystemSphPS(THSystem):
+
+    """This class models heat transfer in a sphere with point heat source at the
+    center and convective heat transfer between the sphere and liquid flow
+    """
+
+    def __init__(self, kappa, components):
+        THSystem.__init__(self, kappa, components)
+
+    def dtempdt(self, component, power, omegas, t_idx):
+        to_ret = 0*units.kelvin/units.second
+        cap = (component.rho(t_idx)*component.cp*component.vol)
+        if component.heatgen:
+            to_ret += self.heatgen(component, power, omegas)/cap
+        for name, d in component.adv.iteritems():
+            Qadv = self.advection(t_out=component.T[t_idx]*2.0 - d['t_in'],
+                                  t_in=d['t_in'],
+                                  m_flow=d['m_flow'],
+                                  cp=d['cp'])
+            to_ret -= Qadv/cap
+            assert Qadv.magnitude >= 0, '''
+            at step %d, %s is heating the system by %f watts???
+            Tin is %f, Tout is %f, Tcomp is %f''' % (
+                t_idx, component.name, Qadv.magnitude, d['t_in'].magnitude,
+                (component.T[t_idx]*2 - d['t_in']).magnitude,
+                component.T[t_idx].magnitude)
+        for interface, d in component.cond.iteritems():
+            env = self.comp_from_name(interface)
+            Qcond = self.conduction(t_b=component.T[t_idx],
+                                    t_env=env.T[t_idx],
+                                    r_b=d['r_b'],
+                                    r_env=d['r_env'],
+                                    k=d['k'])
+            to_ret -= Qcond/cap
+            assert (Qcond*(component.T[t_idx]-env.T[t_idx])).magnitude >= 0, \
+                'conduction from %s to %s, from temp %f to %f is wrong %f' % (
+                    component.name, env.name, component.T[t_idx].magnitude,
+                    env.T[t_idx].magnitude, Qcond.magnitude)
+        for interface, d in component.conv.iteritems():
+            env = self.comp_from_name(interface)
+            Qconv = self.convection(t_b=component.T[t_idx],
+                                    t_env=env.T[t_idx],
+                                    h=d['h'],
+                                    A=d['area'])
+            to_ret -= Qconv/cap
+            assert (Qconv*(component.T[t_idx]-env.T[t_idx])).magnitude >= 0,\
+                'convection from %s to %s, from temp %f to %f is wrong %f' % (
+                    component.name, env.name, component.T[t_idx].magnitude,
+                    env.T[t_idx].magnitude, Qconv.magnitude)
+        return to_ret
+
+    def conduction(self, t_b, t_env, r_b, r_env, k):
+        """
+        heat transfer by conduction(watts)
+        :param t_b: The temperature of the body
```

Comment:
need one blank line before the parameter definitions. sphinx-apidoc syntax requires it to build properly.